File: openadmin/spec/bar_chart.py
# SPDX-FileCopyrightText: 2026 OpenAdmin
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from typing import Literal, NotRequired, TypedDict

from .colors import Color
from .http_methods import HttpMethod
from .icons import Icon
from .json_schema import JsonSchema


class BarChartConfigValue(TypedDict):
    name: NotRequired[str]
    color: NotRequired[Color]
    icon: NotRequired[Icon]


class BarChartComponent(TypedDict):
    type: Literal["bar-chart"]
    id: str
    name: str
    description: str | None
    config: dict[str, BarChartConfigValue] | None
    data_key: str | None
    icon: Icon | None
    color: Color | None
    caption: str | None
    caption_description: str | None
    caption_icon: Icon | None
    refresh: int | None
    method: HttpMethod
    form: JsonSchema | None
    body: JsonSchema | None
    query: JsonSchema | None


type BarChartData = (
    list[dict[str, int | float | str]]
    | list[dict[Literal["data", "value"], int | float | str]]
    | object
)


class BarChartResponce(TypedDict):
    config: NotRequired[dict[str, BarChartConfigValue]]
    icon: NotRequired[Icon]
    color: NotRequired[Color]
    refresh: NotRequired[int | None]
    data: BarChartData


type BarChart = BarChartData | BarChartResponce

File: client/src/schemas/bar-chart.ts
// SPDX-FileCopyrightText: 2026 OpenAdmin
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import { z } from "zod"
import { colorSchema } from "./color"
import { httpMethodSchema } from "./http-method"
import { iconSchema } from "./icon"
import { jsonSchemaSchema } from "./json-schema"

export const barChartConfigValueSchema = z.object({
	name: z.string().optional(),
	color: colorSchema.optional(),
	icon: iconSchema.optional(),
})

export type BarChartConfigValue = z.infer<typeof barChartConfigValueSchema>

export const barChartComponentSchema = z.object({
	type: z.literal("bar-chart"),
	id: z.string(),
	name: z.string(),
	description: z.string().nullable(),
	config: z.record(z.string(), barChartConfigValueSchema).nullable(),
	data_key: z.string().nullable(),
	icon: iconSchema.nullable(),
	color: colorSchema.nullable(),
	caption: z.string().nullable(),
	caption_description: z.string().nullable(),
	caption_icon: iconSchema.nullable(),
	refresh: z.number().nullable(),
	method: httpMethodSchema,
	form: jsonSchemaSchema.nullable(),
	body: jsonSchemaSchema.nullable(),
	query: jsonSchemaSchema.nullable(),
})

export type BarChartComponent = z.infer<typeof barChartComponentSchema>

export const barChartDataSchema = z.union([
	z.array(z.record(z.string(), z.union([z.number(), z.string()]))),
	z.array(z.record(z.enum(["data", "value"]), z.union([z.number(), z.string()]))),
	z.unknown(),
])

export type BarChartData = z.infer<typeof barChartDataSchema>

export const barChartResponseSchema = z.object({
	config: z.record(z.string(), barChartConfigValueSchema).optional(),
	icon: iconSchema.optional(),
	color: colorSchema.optional(),
	refresh: z.number().nullable().optional(),
	data: barChartDataSchema,
})

export type BarChartResponse = z.infer<typeof barChartResponseSchema>

export const barChartSchema = z.union([barChartDataSchema, barChartResponseSchema])

export type BarChart = z.infer<typeof barChartSchema>
